plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go

```diff
@@ -202,150 +114,120 @@ func onAddOrUpdateService(obj interface{}, ipToServiceAndNamespace, serviceAndNa
-		ipToPod.Store(pod.Status.PodIP, pod.Name)
+	}
+	if pod.Status.PodIP != "" {
+		p.ipToPod.Store(pod.Status.PodIP, pod.Name)
 	}
 }
 
-func handlePodUpdate(newPod *corev1.Pod, oldPod *corev1.Pod, ipToPod *sync.Map, deleter Deleter) {
-	if oldPod.Spec.HostNetwork && newPod.Spec.HostNetwork {
-		// Case 1: Both oldPod and newPod are using host network
-		// Here we need to update the host network records accordingly
-		updateHostNetworkRecords(newPod, oldPod, ipToPod, deleter)
-	} else if oldPod.Spec.HostNetwork && !newPod.Spec.HostNetwork {
-		// Case 2: The oldPod was using the host network, but the newPod is not
-		// Here we remove the old host network records and add new PodIP record if it is not empty
-		removeHostNetworkRecords(oldPod, ipToPod, deleter)
-		if newPod.Status.PodIP != "" {
-			ipToPod.Store(newPod.Status.PodIP, newPod.Name)
-		}
-	} else if !oldPod.Spec.HostNetwork && newPod.Spec.HostNetwork {
-		// Case 3: The oldPod was not using the host network, but the newPod is
-		// Here we remove the old PodIP record and add new host network records
-		if oldPod.Status.PodIP != "" {
-			deleter.DeleteWithDelay(ipToPod, oldPod.Status.PodIP)
+func (p *podWatcher) handlePodUpdate(newPod *corev1.Pod, oldPod *corev1.Pod) {
+	// HostNetwork is an immutable field
+	if newPod.Spec.HostNetwork && oldPod.Status.HostIP != newPod.Status.HostIP {
+		if oldPod.Status.HostIP != "" {
+			p.logger.Debug("deleting host ip from cache", zap.String("hostNetwork", oldPod.Status.HostIP))
+			for _, port := range getHostNetworkPorts(oldPod) {
+				p.deleter.DeleteWithDelay(p.ipToPod, oldPod.Status.HostIP+":"+port)
+			}
 		}
-		for _, port := range getHostNetworkPorts(newPod) {
-			ipToPod.Store(newPod.Status.HostIP+":"+port, newPod.Name)
+		if newPod.Status.HostIP != "" {
+			for _, port := range getHostNetworkPorts(newPod) {
+				p.ipToPod.Store(newPod.Status.HostIP+":"+port, newPod.Name)
+			}
 		}
-	} else if !oldPod.Spec.HostNetwork && !newPod.Spec.HostNetwork && oldPod.Status.PodIP != newPod.Status.PodIP {
-		// Case 4: Both oldPod and newPod are not using the host network, but the Pod IPs are different
-		// Here we replace the old PodIP record with the new one
+	}
+	if oldPod.Status.PodIP != newPod.Status.PodIP {
 		if oldPod.Status.PodIP != "" {
-			deleter.DeleteWithDelay(ipToPod, oldPod.Status.PodIP)
+			p.logger.Debug("deleting pod ip from cache", zap.String("podNetwork", oldPod.Status.PodIP))
+			p.deleter.DeleteWithDelay(p.ipToPod, oldPod.Status.PodIP)
 		}
 		if newPod.Status.PodIP != "" {
-			ipToPod.Store(newPod.Status.PodIP, newPod.Name)
+			p.ipToPod.Store(newPod.Status.PodIP, newPod.Name)
 		}
 	}
 }
 
-func onAddOrUpdatePod(newObj, oldObj interface{}, ipToPod, podToWorkloadAndNamespace, workloadAndNamespaceToLabels *sync.Map, workloadPodCount map[string]int, isAdd bool, logger *zap.Logger, deleter Deleter) {
-	pod := newObj.(*corev1.Pod)
-
+func (p *podWatcher) onAddOrUpdatePod(pod, oldPod *corev1.Pod, isAdd bool) {
```

Comment:
Couldn't we determine if it's an add based on the presence of `oldPod`?

Reply:
We can. Updated.